core/endorser/endorser.go

```diff
@@ -458,9 +464,92 @@ func (e *Endorser) ProcessProposalSuccessfullyOrError(up *UnpackedProposal) (*pb
 		Endorsement: endorsement,
 		Payload:     mPrpBytes,
 		Response:    res,
+		Interest:    ccInterest,
 	}, nil
 }
 
+// Using the simulation results, build the ChaincodeInterest structure that the client can pass to the discovery service
+// to get the correct endorsement policy for the chaincode(s) and any collections encountered.
+func (e *Endorser) buildChaincodeInterest(simResult *ledger.TxSimulationResults) (*pb.ChaincodeInterest, error) {
+	// build a structure that collates all the information needed for the chaincode interest:
+	// 1. chaincodes in public RWset that don't have collections
+	// 2. collections in the private write set
+	// 3. collections in the PrivateReads struct
+	chaincodes := map[string]map[string]bool{} // map namespace -> collection -> isRead
+
+	// 1. Add any chaincodes (namespace) from the public RW set
+	if simResult.PubSimulationResults != nil {
+		for _, nsrws := range simResult.PubSimulationResults.GetNsRwset() {
+			if e.Support.IsSysCC(nsrws.Namespace) {
+				// skip system chaincodes
+				continue
+			}
+			chaincodes[nsrws.Namespace] = nil // no collections (maybe added later)
+		}
+	}
+
+	// 2. Add any chaincodes/collections from the private write set
+	if simResult.PvtSimulationResults != nil {
+		// list collections that are read from (in order to populate NoPublicReads in ChaincodeInterest)
+		for _, prws := range simResult.PvtSimulationResults.GetNsPvtRwset() {
+			if e.Support.IsSysCC(prws.Namespace) {
+				// skip system chaincodes
+				continue
+			}
+			// build a map of collections that are read from or written to - value is true if read from, false if only written to
+			collections := map[string]bool{}
+			for _, cprws := range prws.GetCollectionPvtRwset() {
+				collections[cprws.CollectionName] = false
+			}
+			chaincodes[prws.Namespace] = collections
+		}
+	}
+
+	// 3. Add/merge collections from the PrivateReads struct
+	for ns, entry := range simResult.PrivateReads {
+		if cc := chaincodes[ns]; cc == nil {
```

Comment:
- Skip systems chaincode in this block as well
- Add a test that asserts that the system chaincodes are indeed skipped